cocos/audio/src/AudioDecoderOgg.cpp

```diff
@@ -39,21 +39,25 @@ namespace cocos2d {
 
     static size_t ov_fread_r(void* buffer, size_t element_size, size_t element_count, void* handle)
     {
-        return ((FileStream*)handle)->read(buffer, static_cast<uint32_t>(element_size * element_count));
+        auto& fs = *static_cast<std::unique_ptr<FileStream>*>(handle);
+        return fs->read(buffer, static_cast<uint32_t>(element_size * element_count));
     }
 
     static int ov_fseek_r(void * handle, ogg_int64_t offset, int whence)
     {
-        return ((FileStream*)handle)->seek(offset, whence) < 0 ? -1 : 0;
+        auto& fs = *static_cast<std::unique_ptr<FileStream>*>(handle);
+        return fs->seek(offset, whence) < 0 ? -1 : 0;
     }
     
     static long ov_ftell_r(void * handle)
     {
-        return ((FileStream*)handle)->seek(0, SEEK_CUR);
+        auto& fs = *static_cast<std::unique_ptr<FileStream>*>(handle);
+        return fs->tell();
     }
 
     static int ov_fclose_r(void* handle) {
-        ((FileStream*)handle)->close();
+        auto& fs = *static_cast<std::unique_ptr<FileStream>*>(handle);
```

Comment:
Use `ov_open_callbacks(_fileStream.get(), ...) ` for ogg callbacks, and access via `FileStream*` in ogg callbacks may be better.

Reply:
The problem is this `ov_fclose_r`:
```
    static int ov_fclose_r(void* handle) {
        auto& fs = *static_cast<std::unique_ptr<FileStream>*>(handle);
        fs = nullptr;
        return 0;
    }
```

I can change to be like this:
```
    static int ov_fclose_r(void* handle) {
        auto* fs = static_cast<FileStream*>(handle);
        fs->close();
        return 0;
    }
```

The pointer will still be valid until this is called to free the FileStream:
```
    void AudioDecoderOgg::close()
    {
        if (isOpened())
        {
            ov_clear(&_vf);
            _isOpened = false;
            _fileStream.reset();
        }
    }
```

Reply:
Maybe we can make ov_fclose_r empty when the fileStream auto managed by unique_ptr

Reply:
That should work too, so I'll change it so `ov_fclose_r` does nothing.  If any issues come up because of it, we can just add the call to FileStream::close() in there.

Reply:
sounds good

Reply:
or maybe same with FreeType streaming parsing mode is better?
```cpp
       auto fullPath = FileUtils::getInstance()->fullPathForFilename(fontName);
        if (fullPath.empty()) return false;

        auto fs = FileUtils::getInstance()->openFileStream(fullPath, FileStream::Mode::READ);
        if (!fs)
        {
            return false;
        }

        std::unique_ptr<FT_StreamRec> fts(new FT_StreamRec());
        fts->read = ft_stream_read_callback;
        fts->close = ft_stream_close_callback;
        fs->seek(0, SEEK_END);
        fts->size = fs->tell();
        fs->seek(0, SEEK_SET);

        fts->descriptor.pointer = fs.release();
```

Reply:
> or maybe same with FreeType streaming parsing mode is better?

Sorry, I'm not sure what you mean by that.

Reply:
1. Remove std::unique_ptr<FileStream> _fileStream from class AudioDecoderOgg
2. delete stream ta ov_close_r
```
static int ov_fclose_r(void* handle) {
auto* fs = static_cast<FileStream*>(handle);
delete fs;
return 0;
}
```
3. Then the ov_open_callbacks code could be:
```cpp
bool AudioDecoderOgg::open(const std::string& fullPath)
    {
        auto fs = FileUtils::getInstance()->openFileStream(fullPath, FileStream::Mode::READ);
        if (!fs)
        {
            ALOGE("Trouble with ogg(1): %s\n", strerror(errno));
            return false;
        }

        static ov_callbacks OV_CALLBACKS_POSIX = {
               ov_fread_r,
               ov_fseek_r,
               ov_fclose_r,
               ov_ftell_r
        };
              
        if (0 == ov_open_callbacks(fs.release(), &_vf, nullptr, 0, OV_CALLBACKS_POSIX))
        {
            // header
            vorbis_info* vi = ov_info(&_vf, -1);
            _sampleRate = static_cast<uint32_t>(vi->rate);
            _channelCount = vi->channels;
            _bytesPerBlock = vi->channels * sizeof(short);
            _totalFrames = static_cast<uint32_t>(ov_pcm_total(&_vf, -1));
            _isOpened = true;
            return true;
        }
        return false;
    }

```

Reply:
Oh I understand what you mean now, and yes that is a cleaner implementation, reducing the duplication of the pointer storage. I'll modify it now.

Reply:
I'll add `.release()` on the same line so that there isn't a duplicate unique_ptr and raw pointer floating around.  It'll prevent any accidental usage of the unique_ptr after it's been released.

Reply:
it's ok, good job.